rpc/wrtc_client_channel_test.go

```diff
@@ -599,3 +599,88 @@ func TestWebRTCClientChannelCanStopStreamRecvMsg(t *testing.T) {
 
 	<-serverFinished
 }
+
+func TestClientStreamCancel(t *testing.T) {
+	// Tests that clients can cancel server streams over WebRTC.
+	// 1. We set up a server with a stream endpoint and call it with a client.
+	// 2. After the initial request, the client closes its send side of the stream (replicating what happens in a server-side streaming scenario).
+	// 2. 3 messages are sent from the server. After the 3rd message, the client stream is cancelled.
+	// 3. The server should receive a RST_STREAM message and cancel the server context.
+	testutils.SkipUnlessInternet(t)
+	logger := golog.NewTestLogger(t)
+	pc1, pc2, dc1, dc2 := setupWebRTCPeers(t)
+
+	clientCh := newWebRTCClientChannel(pc1, dc1, logger, nil, nil)
+	defer func() {
+		test.That(t, clientCh.Close(), test.ShouldBeNil)
+	}()
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	server := newWebRTCServer(logger)
+	server.RegisterService(&grpc.ServiceDesc{
+		ServiceName: "service_name",
+		Streams: []grpc.StreamDesc{
+			{
+				StreamName: "stream_name",
+				Handler: grpc.StreamHandler(func(srv any, stream grpc.ServerStream) error {
+					defer wg.Done()
+					pongStatus, _ := status.FromError(errors.New("pong"))
+
+					for i := 0; i < 10; i++ {
+						if stream.Context().Err() != nil {
+							return nil
+						}
+						stream.SendMsg(pongStatus.Proto())
+
+						// Using channels is not enough ensure that the client reset
+						// went through in time.
+						// Sleep here to let client process messages and send reset.
+						time.Sleep(10 * time.Millisecond)
+					}
+					// Failure as this means the context was never cancelled.
+					t.Fail()
```

Comment:
[minor]

```suggestion
					t.FailNow()
```

Reply:
FailNow does not work if called in a goroutine (which this is)

Reply:
ah, good to know